lib/easypost/services/shipment.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+require 'set'
+
+# The workhorse of the EasyPost API, a Shipment is made up of a "to" and "from" Address, the Parcel
+# being shipped, and any customs forms required for international deliveries.
+class EasyPost::Services::Shipment < EasyPost::Services::Service
+  MODEL_CLASS = EasyPost::Models::Shipment
+
+  # Create a Shipment.
+  def create(params = {}, with_carbon_offset = false)
+    wrapped_params = {
+      shipment: params,
+      carbon_offset: with_carbon_offset,
+    }
+
+    @client.make_request(:post, 'shipments', MODEL_CLASS, wrapped_params)
+  end
+
+  # Retrieve a Shipment.
+  def retrieve(id)
+    @client.make_request(:get, "shipments/#{id}", MODEL_CLASS)
+  end
+
+  # Retrieve a list of Shipments
+  def all(params = {})
+    response = @client.make_request(:get, 'shipments', MODEL_CLASS, params)
+    response.define_singleton_method(:purchased) { params[:purchased] }
+    response.define_singleton_method(:include_children) { params[:include_children] }
+    response
+  end
+
+  # Get the next page of shipments.
+  def get_next_page(collection, page_size = nil)
+    get_next_page_helper(collection, collection.shipments, 'shipments', MODEL_CLASS, page_size)
+  end
+
+  # Regenerate the rates of a Shipment.
+  def regenerate_rates(id, with_carbon_offset = false)
+    params = { carbon_offset: with_carbon_offset }
+
+    @client.make_request(:post, "shipments/#{id}/rerate", MODEL_CLASS, params)
+  end
+
+  # Get the SmartRates of a Shipment.
+  def get_smartrates(id)
+    @client.make_request(:get, "shipments/#{id}/smartrate", MODEL_CLASS).result || []
+  end
+
+  # Buy a Shipment.
+  def buy(id, params = {}, with_carbon_offset = false, end_shipper_id = nil)
+    if params.instance_of?(EasyPost::Models::Rate)
+      params = { rate: params.clone }
+    end
+
+    params[:carbon_offset] = params[:with_carbon_offset] || with_carbon_offset
+    params.delete(:with_carbon_offset)
+
+    params[:end_shipper_id] = end_shipper_id if end_shipper_id
+
+    @client.make_request(:post, "shipments/#{id}/buy", MODEL_CLASS, params)
```

Comment:
question: I assume that since client is an instance variable, it will not be accessible to users outside correct? Meaning they couldn't call the client object of this service? Or could they?